### PR TITLE
apiserver expose a handler to indicate whether admission webhook configurations are honored

### LINKED ISSUE #54712: Admission webhook configuration API should report if the configuration is effective
The webhook [api](https://github.com/kubernetes/kubernetes/blob/v1.9.0-alpha.1/staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go#L152) doesn't tell if a webhook configuration is honored by the kube-apiserver or the extension apiservers. We should support it.

The problem is less critical for initializers, because users can observe objects's `metadata.initializers` to check if the initializer configuration is honored

cc @lavalamp @cheftako @kubernetes/sig-api-machinery-misc

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
expose a handler to provide a way to know whether admission configurations are honored.

fixes #54712

**Release note**:

```release-note
 Apiservers expose their admission webhook configs via `/configz` to let users know whether their configurations are honored.
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/admission/configuration/validating_webhook_manager.go [hunk 1]
```diff
@@ -57,6 +61,13 @@ func NewValidatingWebhookConfigurationManager(f informers.SharedInformerFactory)
 		DeleteFunc: func(_ interface{}) { manager.updateConfiguration() },
 	})
 
+	// setup /configz endpoint
+	if cz, err := configz.New("validatingwebhookconfigurations"); err == nil {
+		manager.configz = cz
+	} else {
+		glog.Errorf("unable to register configz: %c", err)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/configuration/validating_webhook_manager.go]: %c?

File: staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go [hunk 2]
```diff
@@ -57,6 +61,13 @@ func NewMutatingWebhookConfigurationManager(f informers.SharedInformerFactory) g
 		DeleteFunc: func(_ interface{}) { manager.updateConfiguration() },
 	})
 
+	// setup /configz endpoint
+	if cz, err := configz.New("mutatingwebhookconfigurations"); err == nil {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go]: isn't the "configurations" postfix implicit?

File: staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go [hunk 3]
```diff
@@ -57,6 +61,13 @@ func NewMutatingWebhookConfigurationManager(f informers.SharedInformerFactory) g
...
+		manager.configz = cz
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go]: Do we open up another API this way? Do we document the compatibility guarantees for that?

File: staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go [hunk 4]
```diff
@@ -76,6 +87,9 @@ func (m *mutatingWebhookConfigurationManager) updateConfiguration() {
 		return
 	}
 	m.configuration.Store(mergeMutatingWebhookConfigurations(configurations))
+	if m.configz != nil {
+		m.configz.Set(configurations)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go]: Set is thread safe?

File: staging/src/k8s.io/apiserver/pkg/server/config_test.go [hunk 5]
```diff
@@ -98,6 +98,7 @@ func TestNewWithDelegate(t *testing.T) {
   "paths": [
     "/apis",
     "/bar",
+    "/configz",
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config_test.go]: is this endpoint protected by default? If not, isn't the admission chain security sensitive?

Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/server/config_test.go]: >is this endpoint protected by default?

yes

File: staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go (hunk 4 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go]: yes

File: staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go (hunk 3 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go]: No, this is one of the `/configz`  configurations, which is a map.

**Comment by k82cn** [comment]
/approve  for scheduler part

And removed sig-scheduling labels :)

**Comment by sttts** [comment]
We have put the topic of coordination of dynamic config onto the agenda of the sig-api-machinery meeting. We need an answer how to proceed with this topic in a general context, not only with admission webhooks. Holding this until we know better what we want.

/hold

**Comment by lavalamp** [comment]
This needs to be part of the regular API (i.e., status in the webhook config object), not a special endpoint. Judging from the description of this PR I'm not going to want to do it this way.

**Comment by hzxuzhonghu** [comment]
@lavalamp Sure， will wait for a good design.

**Comment by lavalamp** [comment]
Thanks, it's on the agenda for next week's sig meeting.

**Pull request closed without merging**